File: jpablog/src/main/java/com/example/jpablog/repository/UserRepository.java
package com.example.jpablog.repository;

import com.example.jpablog.model.User;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.stereotype.Repository;

@Repository // 생략가능
public interface UserRepository extends JpaRepository<User, Long> {








//    // select * from user where username =? and password =?;
//    User findByUsernameAndPassword(String username, String password);

//    @Query(value = "select * from user where username=?1 and password=?2", nativeQuery = true)
//    User login(String username, String password);
}

File: chat-server/chat-persistence/src/main/kotlin/com/chat/persistence/repository/UserRepository.kt
package com.chat.persistence.repository

import com.chat.domain.model.User
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Modifying
import org.springframework.data.jpa.repository.Query
import org.springframework.stereotype.Repository
import java.time.LocalDateTime

@Repository
interface UserRepository : JpaRepository<User, Long> {

    fun findByUsername(username: String): User?
    fun existsByUsername(username: String): Boolean

    @Modifying
    @Query("UPDATE User u SET u.lastSeenAt = :lastSeenAt WHERE u.id = :userId")
    fun updateLastSeenAt(userId: Long, lastSeenAt: LocalDateTime)

    @Query("SELECT u FROM User u WHERE " +
            "LOWER(u.username) LIKE LOWER(CONCAT('%', :query, '%')) OR " +
            "LOWER(u.displayName) LIKE LOWER(CONCAT('%', :query, '%'))")
    fun searchUsers(query: String, pageable: Pageable): Page<User>
}